docs/design.md

```diff
@@ -967,6 +864,92 @@ spare capacity is chosen in the same datacenter and a special-case split
+as it is simpler to reason about and includes a reference implementation
+covering important details. Every write to replicas is logged twice.
+Once to RocksDB’s internal log and once to levedb itself as part of the
+Raft consensus log.
+[ePaxos](https://www.cs.cmu.edu/~dga/papers/epaxos-sosp2013.pdf) has
+promising performance characteristics for WAN-distributed replicas, but
+it does not guarantee a consistent ordering between replicas.
+
+Raft elects a relatively long-lived leader which must be involved to
+propose commands. It heartbeats followers periodically and keeps their logs
+replicated. In the absence of heartbeats, followers become candidates
+after randomized election timeouts and proceed to hold new leader
+elections. Cockroach weights random timeouts such that the replicas with
+shorter round trip times to peers are more likely to hold elections
+first (not implemented yet). Only the Raft leader may propose commands;
+followers will simply relay commands to the last known leader.
+
+Our Raft implementation was developed together with CoreOS, but adds an extra
+layer of optimization to account for the fact that a single Node may have
+millions of consensus groups (one for each Range). Areas of optimization
+are chiefly coalesced heartbeats (so that the number of nodes dictates the
+number of heartbeats as opposed to the much larger number of ranges) and
+batch processing of requests.
+Future optimizations may include two-phase elections and quiescent ranges
+(i.e. stopping traffic completely for inactive ranges).
+
+# Range Leadership (Leader Leases)
+
+As outlined in the Raft section, the replicas of a Range are organized as
+a Raft group and execute commands from their shared commit log. Going
+through Raft is an expensive operation though, and there are tasks which
+should only be carried out by a single replica at a time (as opposed to
+all of them).
+For these reasons, Cockroach introduces the concept of a 
+**Range Leader**: This is a lease held for a slice of (database, i.e. hybrid
+logical) time and is established by committing a special log entry through Raft
+containing the interval the leadership is going to be active on, along with the
+Node:RaftID combination that uniquely describes the requesting replica.
+Reads and writes must generally be addressed to the replica holding the lease;
+if none does, any replica may be addressed, causing it to try to obtain the
+lease synchronously. Requests received by a non-leader (for the HLC timestamp
+specified in the request's header) fail with an error pointing at the replica's
+last known leader. These requests are retried transparently with the updated
+leader by the gateway node and never reach the client.
+
+The replica holding the lease is in charge of handling Range-specific
+maintenance tasks such as
+
+* gossiping the sentinel and/or first range information
+* splitting, merging and rebalancing
```

Comment:
When a node attempts to merge two ranges, it seems that we currently checks only the leader lease of the left-side range. Do we also need to acquire the leader lease of the subsumed range?